### PR TITLE
Tweak a localization to remove any assumption on "+" localization

### PR DESCRIPTION
**Pull request by mokagio**
See discussion at https://github.com/wordpress-mobile/WordPress-iOS/pull/19472#discussion_r996700821.

I opened a dedicated PR for this instead of committing directly to the release branch because the change is big and I didn't want to let a mistake land directly in the release branch.

Notice https://github.com/wordpress-mobile/WordPress-iOS/pull/19473/commits/daec387a70218ba91c4799c1cb2e54193dc44329 which successfully updates the `.strings` via the `generate_strings_file_for_glotpress` lane. I made this as part of this PR to verify my changed was correct, as far as our tooling, or rather `genstring`, goes.

---

@staskus you have more context on this than I do. I think I implemented a 1-1 version of the end result that https://github.com/wordpress-mobile/WordPress-iOS/commit/e3c3e41637a5319dd2577a3bba9be88a1ed6349a would have produced. Still, why don't we show a minus sign, `-`, when the difference is less than 0?

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift
Comment (AliSoftware) [inline WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift]:
```suggestion
                "insights.visitorsLineChartCell.differenceLabelPositiveWithPercentage",
```

File: WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift
Comment (AliSoftware) [inline WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift]:
```suggestion
                "insights.visitorsLineChartCell.differenceLabelPositiveWithoutPercentage",
```

File: WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift
Comment (AliSoftware) [inline WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift]:
```suggestion
                "insights.visitorsLineChartCell.differenceLabelNonPositiveWithPercentage",
```

Or (depending on @staskus 's reply to your answer in the PR description about why we don't use negative sign in the copy):

```suggestion
                "insights.visitorsLineChartCell.differenceLabelNegativeWithPercentage",
```

**Review by AliSoftware — DISMISSED** [review]

**Comment by staskus** [comment]
@mokagio 

Thank you for the fixes! I'll review it now.

> Still, why don't we show a minus sign, -, when the difference is less than 0?

We do show negative signs. Negative numbers, both absolute and percentages appear with a "-" automatically when they're negative numbers. So it wasn't explicitly included in the localization.

However, to be explicit from the localization POV, we could pass `abs(...)` numbers to a formatting function and request a localization with a minus (or an equivalent localized) sign.

File: WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift
Comment (staskus) [inline WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift]: So yes, if we want to be explicit about the localization negative case, then we should handle cases when `difference == 0` and `difference < 0` differently:

1. `difference == 0` no sign
2. `difference < 0` negative sign in a localization, pass `abs(difference).abbreviatedString()`

File: WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift
Comment (staskus) [inline WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift:96]: Same here, we could have 2 different cases if `difference == 0` or `difference < 0`

**Comment by mokagio** [comment]
> However, to be explicit from the localization POV, we could pass `abs(...)` numbers to a formatting function and request a localization with a minus (or an equivalent localized) sign.

I think that would be the best course of action, consistent with the rationale applied for the `+` sign.

It's too late in my day. If no one gets to this before me, I'll look at this ASAP tomorrow.

**Comment by AliSoftware** [comment]
> However, to be explicit from the localization POV, we could pass `abs(...)` numbers to a formatting function and request a localization with a minus (or an equivalent localized) sign.

Actually, I'm not sure if this would be a better approach compared to using a `NSNumberFormatter` and let it do the proper formatting depending on the user's locale for us automatically. In face, if we configure our `NSNumberFormatter` property to even include the `+` sign if the value is positive (I think the API has an option for it?), then we'd be back to just needing only 2 strings to localize (with or without the %), and let the whole formatting logic to the system instead of letting the polyglots assume responsibility for it.

Actually, while we're at it, we should even use an `NumberFormatter` for the percentage as well, so that it uses the right `numberFormatter.percentSymbol` automatically depending on the formatter's `locale` (assuming we use `numberStyle = .percent` on it)… and if we do so, we might not even need any special localization anymore?

---

<details><summary>For example, I think this should do it without even requiring manual (<tt>NSLocalizedString</tt>) localization at all, just letting it all to the system to handle all cases</summary>

```swift
    var differenceLabel: String {
        var string: String
        
        let formatter = NumberFormatter()
        formatter.positivePrefix = formatter.plusSign // Enforce + sign if value is positive
        string = formatter.string(from: difference as NSNumber) ?? ""
        
        if differencePercent != 0 {
            formatter.numberStyle = .percent
            formatter.positivePrefix = "" // Reset the + prefix for the percentage value. Or should we?
            string += formatter.string(from: differencePercent as NSNumber).map { " (\($0))" } ?? ""
        }
        return string
    }
```
</details>

---

[EDIT] Oh wait, doing so won't output `5.7k` for `difference = 5_700`, but would output `5700` instead. So we would lose the benefit of `abbreviatedString`…

<details><summary>Ok, so if you want the <tt>k</tt> you might rely on <a href="https://developer.apple.com/documentation/foundation/measurementformatter"><tt>MeasurementFormatter</tt></a> instead…</summary>

…and a [dimensionless `Unit` subclass](https://developer.apple.com/documentation/foundation/unit)

```swift
@objc class VisitCount: Unit {
    override init(symbol: String) { super.init(symbol: symbol) }
    @available(*, unavailable)
    required init(coder: NSCoder) { fatalError("Unimplemented") }
    
    static func measurement(from count: Int) -> Measurement<VisitCount> {
        switch abs(count) {
        case 0..<1_000:
            return Measurement(value: Double(count), unit: VisitCount(symbol: ""))
        case ..<1_000_000:
            let symbol = NSLocalizedString("visit.count.prefix.thousand", value: "k")
            return Measurement(value: Double(count/1_000), unit: VisitCount(symbol: symbol)
        default:
            let symbol = NSLocalizedString("visit.count.prefix.millions", value: "M")
            return Measurement(value: Double(count/1_000_000), unit: VisitCount(symbol: symbol))
        }
    }
}

struct StatsSegmentedControlData {
    …
    let difference: Int
    let differencePercent: Int
    …
    
    var differenceLabel: String {
        var string: String

        let nf = NumberFormatter()
        nf.positivePrefix = nf.plusSign // Enforce + sign if value is positive
        
        let mf = MeasurementFormatter()
        mf.numberFormatter = nf
        let differenceMeasurement = VisitCount.measurement(from: difference)
        string = mf.string(from: differenceMeasurement)
        
        if differencePercent != 0 {
            nf.numberStyle = .percent
//            nf.positivePrefix = "" // Reset the + prefix for the percentage value. Or should we?
            string += nf.string(from: Double(differencePercent)/100.0 as NSNumber).map { " (\($0))" } ?? ""
        }
        return string
    }
}
```

</details>

But going with that solution might start to go a bit astray from the current implementation and kind of re-inventing the `abbreviatedString()` utility… so that might bring us a bit too much changes for a PR targeting the `release/` branch, and we might want to keep that for a future PR for next version instead.

If our `abbreviatedString()`'s implementation is already localized, then maybe just using `NumberFormatter().plusSign` as a string prefix if value>0, and using a `NumberFormatter` for the percent part when it's needed, can be enough after all, at least for this sprint, and thus not require additional localization of all the cases?

**Comment by staskus** [comment]
@AliSoftware 

> Actually, I'm not sure if this would be a better approach compared to using a NSNumberFormatter and let it do the proper formatting depending on the user's locale for us automatically.

Thanks for putting your thought into it 👍

I agree, handling all these additional cases manually seems like a stretch at the moment, especially given that a lot of the localization logic is already built in.

We could:
- When percentages exist, keep the original `NSLocalizedString(...)` since it was already localized before the release, and improve it with `formatter.plusSign`
- When percentages don't exist, simple return `formatter.plusSign` + `difference.abbreviatedString()`

<details>
  <summary>Working code example</summary>

```
    var differenceLabel: String {
        // We want to show something like "+1.2K (+5%)" if we have a percentage difference and "1.2K" if we don't.
        // Negative cases automatically appear with a negative sign "-1.2K (-5%)" by using abbreviatedString()
        let formatter = NumberFormatter()
        formatter.locale = .current
        let plusSign = difference < 0 ? "" : "\(formatter.plusSign ?? "")"
        
        if differencePercent != 0 {
            let stringFormat = NSLocalizedString("%@%@ (%@%%)", comment: "Difference label for Insights Overview stat, indicating change from previous period. Ex: +99.9K(5%)")
            return String.localizedStringWithFormat(stringFormat,
                    plusSign,
                    difference.abbreviatedString(),
                    differencePercent.abbreviatedString())
        } else {
            return "\(plusSign)\(difference.abbreviatedString())"
        }
    }
```
</details>

**Comment by AliSoftware** [comment]
Sounds like a good compromise.

Only change I'd make to your code example is to use positional placeholders for the `NSLocalizedString` — i.e. `%1$@%2$@ (%3$@%%)` instead of `%@%@ (%@%%)` — and provide a more detailed translator comment to explain what each of the `%1$@`, `%2%@` and %3$@` placeholders will be replaced with 🙂

[See also internal ref pbMoDN-1Df-p2 for more documentation/tips about our suggested practices about localization and positional placeholder]

File: WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift
Comment (mokagio) [inline WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift:87]: I based this off @staskus [suggestion](https://github.com/wordpress-mobile/WordPress-iOS/pull/19473#issuecomment-1281295334) but I decided to use a localized string in the case without the percentage as well, so that translators can apply the appropriate RTL.

I'm not actually sure whether that's necessary, though. 

Maybe we should lean into `NumberFormatter` more for this? There's definitely rooms for improvement and internal refactors. The new tests added in this diff will help us with that in the future.

Comment (staskus) [reply WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift:87]: Thanks for the fix. 👍 I think the solution makes sense. We improved it by using numbered positions, localized plus signs, and most importantly, having localized strings static and not dynamic.

**Review by staskus — APPROVED** [review]
Thank you for the work and additional tests 🙏 

LGTM, I reviewed the code and tested the solution.

**Comment by mokagio** [comment]
Dismissed Olivier's review because I did my best to address the existing concerns, have received another approval, and want to get the code freeze merged into `trunk` (for which this PR is a required step) ASAP.

Proceeding to merge now.

If we missed anything, we can always issue a new round of localizations.

**Pull request merged**
